src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -129,17 +129,43 @@ public void close() {
             }
         }
     }
+    
+    private void ensureHasNext() {
+        if (!currentBlock.hasNext()) {
+            // Rethrow any exceptions raised on the background thread
+            // at the point the exception would have been encountered
+            // if we had performed synchronous iteration
+            raiseBackgroundThreadException();
```

Comment:
This works because once an exception is encountered, the read-ahead block is returned immediately, regardless of the number of records that were added.